cms/lib/xblock/authoring_mixin.py

```diff
@@ -51,3 +56,111 @@ def visibility_view(self, _context=None):
+        Called right *after* the block is written to the DB. Can be used, e.g., to spin up followup tasks.
+
+        By default, is a no-op. Can be overriden in subclasses.
+        """
+
+    def studio_duplicate(
+        self,
+        parent_usage_key,
+        duplicate_source_usage_key,
+        user,
+        store,
+        dest_usage_key=None,
+        display_name=None,
+        shallow=False,
+        is_child=False,
+    ):
+        """
+        Duplicate an existing xblock as a child of the supplied parent_usage_key. You can
+        optionally specify what usage key the new duplicate block will use via dest_usage_key.
+
+        If shallow is True, does not copy children.
+        """
+        from cms.djangoapps.contentstore.utils import gather_block_attributes, load_services_for_studio
+
+        if not dest_usage_key:
+            # Change the blockID to be unique.
+            dest_usage_key = self.location.replace(name=uuid4().hex)
+
+        category = dest_usage_key.block_type
+
+        duplicate_metadata, asides_to_create = gather_block_attributes(
+            self,
+            display_name=display_name,
+            is_child=is_child,
+        )
+
+        dest_block = store.create_item(
+            user.id,
+            dest_usage_key.course_key,
+            dest_usage_key.block_type,
+            block_id=dest_usage_key.block_id,
+            definition_data=self.get_explicitly_set_fields_by_scope(Scope.content),
+            metadata=duplicate_metadata,
+            runtime=self.runtime,
+            asides=asides_to_create,
+        )
+
+        # Allow an XBlock to do anything fancy it may need to when duplicated from another block.
+        load_services_for_studio(self.runtime, user)
+        dest_block.studio_post_duplicate(self, store, user, shallow=shallow)
+        # pylint: disable=protected-access
+        if "detached" not in self.runtime.load_block_type(category)._class_tags:
+            parent = store.get_item(parent_usage_key)
+            # If source was already a child of the parent, add duplicate immediately afterward.
+            # Otherwise, add child to end.
+            if self.location in parent.children:
+                source_index = parent.children.index(self.location)
+                parent.children.insert(source_index + 1, dest_block.location)
+            else:
+                parent.children.append(dest_block.location)
+            store.update_item(parent, user.id)
+
+        # .. event_implemented_name: XBLOCK_DUPLICATED
+        XBLOCK_DUPLICATED.send_event(
+            time=datetime.now(timezone.utc),
+            xblock_info=DuplicatedXBlockData(
+                usage_key=dest_block.location,
+                block_type=dest_block.location.block_type,
+                source_usage_key=duplicate_source_usage_key,
+            ),
+        )
+
+        return dest_block.location
+
+    def studio_post_duplicate(
+        self,
+        source_block,
+        store,
```

Comment:
If we're changing the API of `studio_post_duplicate`, I would love to find a way to remove `store`, since it's specific to modulestore. The `store` parameter wasn't actually used prior to this PR.

Passing `user` also seems a bit redundant as the user is available via the `user` service, as an ID via `self.scope_ids.user_id`, and via `self.runtime.user` (non-standard but works in all Open edX runtimes).

Reply:
i agree, i really wanted to not inject the store dependency but this was the only way i figured i could keep the exact same behavior/code as before. The API is very recent and I doubt anyone uses it considering it's also internal.

I'll wait until @kdmccormick comes back to see what we can do to improve this API.